SPIFlashLogger.class.nut

```diff
@@ -76,53 +114,198 @@ class SPIFlashLogger {
+    //   start   - The first byte used by the logger
+    //   end     - The last byte used by the logger
+    //   sectors - The number of sectors allocated to the logger
+    //   sectorSize - The size of sectors in bytes
     function dimensions() {
         return { "size": _size, "len": _len, "start": _start, "end": _end, "sectors": _sectors, "sector_size": SPIFLASHLOGGER_SECTOR_SIZE }
     }
 
+    // Writes any serializable object to the memory allocated
+    // for the SPIFlashLogger. If the memory is full,
+    // the logger begins overwriting the oldest entries.
+    // If the provided object can not be serialized,
+    // the exception is thrown by the underlying serializer class.
+    //
+    // Parameters:
+    //    object - an object to serialize and save
     function write(object) {
         // Check of the object will fit
-        local obj_len = _serializer.sizeof(object, SPIFLASHLOGGER_OBJECT_MARKER);
-        if (obj_len > _max_data) throw "Cannot store objects larger than alloted memory."
+        local objLength = _serializer.sizeof(object, SPIFLASHLOGGER_OBJECT_MARKER);
+        if (objLength > _maxData) throw "Cannot store objects larger than alloted memory."
 
         // Serialize the object
         local obj = _serializer.serialize(object, SPIFLASHLOGGER_OBJECT_MARKER);
 
         _enable();
 
         // Write one sector at a time with the metadata attached
-        local obj_pos = 0;
-        local obj_remaining = obj_len;
+        local objPos = 0;
+        local objRemaining = objLength;
         do {
 
             // How far are we from the end of the sector
-            if (_at_pos < SPIFLASHLOGGER_SECTOR_META_SIZE) _at_pos = SPIFLASHLOGGER_SECTOR_META_SIZE;
-            local sec_remaining = SPIFLASHLOGGER_SECTOR_SIZE - _at_pos;
-            if (obj_remaining < sec_remaining) sec_remaining = obj_remaining;
+            if (_atPos < SPIFLASHLOGGER_SECTOR_METADATA_SIZE) _atPos = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+            local secRemaining = SPIFLASHLOGGER_SECTOR_SIZE - _atPos;
 
             // We are too close to the end of the sector, skip to the next sector
-            if (obj_pos == 0 && sec_remaining < SPIFLASHLOGGER_OBJECT_MIN_SIZE) {
-                _at_sec = (_at_sec + 1) % _sectors;
-                _at_pos = SPIFLASHLOGGER_SECTOR_META_SIZE;
+            if (objPos == 0 && secRemaining < SPIFLASHLOGGER_OBJECT_MIN_SIZE) {
+                secRemaining = SPIFLASHLOGGER_SECTOR_BODY_SIZE;
+                _atSec = (_atSec + 1) % _sectors;
+                _atPos = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+            }
+            // Handle overflow use-case for a one-sector logger and multiple sectors logger
+            // This check intended to prevent start code overwriting
+            if ((_sectors == 1 && objPos == 0 && objRemaining > secRemaining)
+                || (_sectors > 1
+                    && _atPos > SPIFLASHLOGGER_SECTOR_METADATA_SIZE
```

Comment:
why is the check for `_atPos > SPIFLASHLOGGER_SECTOR_METADATA_SIZE` really needed?